src/AccessControl/cAccessControl.c

```diff
@@ -1847,13 +1847,16 @@ c_rolesForPermissionOn(PyObject *perm, PyObject *object,
   Py_INCREF(r);
 
   /*
-    while 1:
+    while True:
   */
   while (1)
     {
       /*
-        if hasattr(object, n):
+        try:
             roles = getattr(object, n)
+        except (AttributeError, zExceptions_Unauthorized):
+            pass
+        else:
       */
       PyObject *roles = PyObject_GetAttr(object, n);
       if (roles != NULL)
```

Comment:
This still gives us a difference for the Python 3 and C implementations: while the Python 3 implementation only ignores `AttributeError` and `Unauthorized`, the C implementation ignores all exceptions (like Python 2 did).

I am not sure whether this has a practical effect but I think the Python 3 and C implementation should behave the same way (for consistency).

Reply:
We should change the C implementation: ignoring all exceptions may misinterpret temporary exceptions (like e.g. `DatabaseReadConflict`).

Reply:
Thanks, I was also starting to think the same. This still matches the python2 behavior of `hasattr`, but the python3 behavior is better.

I'll try to take care of this (but maybe not right now)

Reply:
C implementation was changed as discussed here